### PR TITLE
fix build failure on windows in android

### PR DESCRIPTION
**Pull request by FouadMagdy01**
## Summary:

This pull request addresses a CMake configuration issue where an invalid escape character in file paths caused the build process to fail. Specifically, it resolves the issue in the React Native CMake configuration file where the path separator was incorrectly handled, leading to an error in the build system.

the issue is in [This Issue](https://github.com/expo/expo/issues/32955) and [This](https://github.com/expo/expo/issues/32957)

## Changelog:

[INTERNAL] [FIXED] - Corrected invalid escape character in CMake path handling

## Test Plan:

To test the changes, I performed the following steps:

1. Cloned the repository and checked out the `fix-cmake-invalid-escape-character` branch.
2. Ran the CMake build on a Windows environment where the issue was previously occurring.
3. Verified that the build process completed successfully without the "invalid character escape" error.
4. Ensured that the path handling now works correctly in CMake on Windows platforms.

### CODE REVIEW

File: packages/react-native/ReactAndroid/cmake-utils/ReactNative-application.cmake [hunk 1]
```diff
@@ -26,45 +26,47 @@ set(REACT_COMMON_DIR ${REACT_ANDROID_DIR}/../ReactCommon)
 # If you have ccache installed, we're going to honor it.
 find_program(CCACHE_FOUND ccache)
 if(CCACHE_FOUND)
-  set_property(GLOBAL PROPERTY RULE_LAUNCH_COMPILE ccache)
-  set_property(GLOBAL PROPERTY RULE_LAUNCH_LINK ccache)
+    set_property(GLOBAL PROPERTY RULE_LAUNCH_COMPILE ccache)
+    set_property(GLOBAL PROPERTY RULE_LAUNCH_LINK ccache)
```
Comment (cipolleschi) [inline packages/react-native/ReactAndroid/cmake-utils/ReactNative-application.cmake:30]: can you revert this change? A good practice is to try and minimize the changes between the two files and this formatting is not necessary.
Plus it will probably requires us to reformatting because of different internal formatting setting.

File: packages/react-native/ReactAndroid/cmake-utils/ReactNative-application.cmake [hunk 2]
```diff
@@ -26,45 +26,47 @@ set(REACT_COMMON_DIR ${REACT_ANDROID_DIR}/../ReactCommon)
...
+    string(REPLACE "\\" "/" BUILD_DIR ${BUILD_DIR})
+    string(REPLACE "\\" "/" REACT_ANDROID_DIR ${REACT_ANDROID_DIR})
 endif()
 
-if (PROJECT_ROOT_DIR)
-# This empty `if` is just to silence a CMake warning and make sure the `PROJECT_ROOT_DIR`
-# variable is defined if user need to access it.
-endif ()
-
+# Glob input source files
```
Comment (cipolleschi) [inline packages/react-native/ReactAndroid/cmake-utils/ReactNative-application.cmake:40]: Why are we removing this `if`?

File: packages/react-native/ReactAndroid/cmake-utils/ReactNative-application.cmake [hunk 3]
```diff
@@ -26,45 +26,47 @@ set(REACT_COMMON_DIR ${REACT_ANDROID_DIR}/../ReactCommon)
...
-                # This can be removed for React Native 0.73
-                -Wno-error=cpp
-                -fexceptions
-                -frtti
-                -std=c++20
-                -DLOG_TAG=\"ReactNative\"
-                -DFOLLY_NO_CONFIG=1
+        -Wall
+        -Werror
+        # Suppress cpp #error and #warning to prevent build failures
```
Comment (cipolleschi) [inline packages/react-native/ReactAndroid/cmake-utils/ReactNative-application.cmake:63]: could you put back the comment that got removed?

File: packages/react-native/ReactAndroid/cmake-utils/ReactNative-application.cmake [hunk 4]
```diff
@@ -81,39 +83,36 @@ target_link_libraries(${CMAKE_PROJECT_NAME}
...
 target_compile_options(common_flags INTERFACE ${folly_FLAGS})
 
-# If project is on RN CLI v9, then we can use the following lines to link against the autolinked 3rd party libraries.
-if(EXISTS ${PROJECT_BUILD_DIR}/generated/autolinking/src/main/jni/Android-autolinking.cmake)
-        include(${PROJECT_BUILD_DIR}/generated/autolinking/src/main/jni/Android-autolinking.cmake)
-        target_link_libraries(${CMAKE_PROJECT_NAME} ${AUTOLINKED_LIBRARIES})
-        foreach(autolinked_library ${AUTOLINKED_LIBRARIES})
-            target_link_libraries(${autolinked_library} common_flags)
-        endforeach()
+# Autolinked libraries if available
```
Comment (cipolleschi) [inline packages/react-native/ReactAndroid/cmake-utils/ReactNative-application.cmake:90]: Could you put back the removed comment?

File: packages/react-native/ReactAndroid/cmake-utils/ReactNative-application.cmake [hunk 5]
```diff
@@ -81,39 +83,36 @@ target_link_libraries(${CMAKE_PROJECT_NAME}
...
+if(EXISTS ${BUILD_DIR}/generated/autolinking/src/main/jni/Android-autolinking.cmake)
+    include(${BUILD_DIR}/generated/autolinking/src/main/jni/Android-autolinking.cmake)
```
Comment (cipolleschi) [inline packages/react-native/ReactAndroid/cmake-utils/ReactNative-application.cmake:92]: Why are we moving from `PROJECT_BUILD_DIR` to just `BUILD_DIR`? Which issue does this solve?

File: packages/react-native/ReactAndroid/cmake-utils/ReactNative-application.cmake [hunk 6]
```diff
@@ -81,39 +83,36 @@ target_link_libraries(${CMAKE_PROJECT_NAME}
...
+    target_link_libraries(${CMAKE_PROJECT_NAME} ${AUTOLINKED_LIBRARIES})
+    foreach(autolinked_library ${AUTOLINKED_LIBRARIES})
+        target_link_libraries(${autolinked_library} common_flags)
+    endforeach()
```
Comment (cipolleschi) [inline packages/react-native/ReactAndroid/cmake-utils/ReactNative-application.cmake:96]: Could you please revert the formatting?

File: packages/react-native/ReactAndroid/cmake-utils/ReactNative-application.cmake [hunk 7]
```diff
@@ -81,39 +83,36 @@ target_link_libraries(${CMAKE_PROJECT_NAME}
...
-        # local app modules can properly be linked.
-        string(REGEX REPLACE "react_codegen_" "" APP_CODEGEN_HEADER "${APP_CODEGEN_TARGET}")
-        target_compile_options(${CMAKE_PROJECT_NAME}
-                PRIVATE
-                -DREACT_NATIVE_APP_CODEGEN_HEADER="${APP_CODEGEN_HEADER}.h"
-                -DREACT_NATIVE_APP_COMPONENT_DESCRIPTORS_HEADER="react/renderer/components/${APP_CODEGEN_HEADER}/ComponentDescriptors.h"
-                -DREACT_NATIVE_APP_COMPONENT_REGISTRATION=${APP_CODEGEN_HEADER}_registerComponentDescriptorsFromCodegen
-                -DREACT_NATIVE_APP_MODULE_PROVIDER=${APP_CODEGEN_HEADER}_ModuleProvider
-        )
+# Link and build the generated library for codegen if available
```
Comment (cipolleschi) [inline packages/react-native/ReactAndroid/cmake-utils/ReactNative-application.cmake:99]: Could you please keep the old comment?

File: packages/react-native/ReactAndroid/cmake-utils/ReactNative-application.cmake [hunk 8]
```diff
@@ -81,39 +83,36 @@ target_link_libraries(${CMAKE_PROJECT_NAME}
...
+if(EXISTS ${BUILD_DIR}/generated/source/codegen/jni/CMakeLists.txt)
+    add_subdirectory(${BUILD_DIR}/generated/source/codegen/jni/ codegen_app_build)
+    get_property(APP_CODEGEN_TARGET DIRECTORY ${BUILD_DIR}/generated/source/codegen/jni/ PROPERTY BUILDSYSTEM_TARGETS)
+    target_link_libraries(${CMAKE_PROJECT_NAME} ${APP_CODEGEN_TARGET})
+    target_link_libraries(${APP_CODEGEN_TARGET} common_flags)
+
+    # Pass generated header and module provider to OnLoad.cpp
```
Comment (cipolleschi) [inline packages/react-native/ReactAndroid/cmake-utils/ReactNative-application.cmake:106]: Keep the old comment please

File: packages/react-native/ReactAndroid/cmake-utils/ReactNative-application.cmake [hunk 9]
```diff
@@ -81,39 +83,36 @@ target_link_libraries(${CMAKE_PROJECT_NAME}
...
+            -DREACT_NATIVE_APP_CODEGEN_HEADER="${APP_CODEGEN_HEADER}.h"
+            -DREACT_NATIVE_APP_COMPONENT_DESCRIPTORS_HEADER="react/renderer/components/${APP_CODEGEN_HEADER}/ComponentDescriptors.h"
+            -DREACT_NATIVE_APP_COMPONENT_REGISTRATION=${APP_CODEGEN_HEADER}_registerComponentDescriptorsFromCodegen
+            -DREACT_NATIVE_APP_MODULE_PROVIDER=${APP_CODEGEN_HEADER}_ModuleProvider
+    )
 endif()
 
-# We set REACTNATIVE_MERGED_SO so libraries/apps can selectively decide to depend on either libreactnative.so
-# or link against a old prefab target (this is needed for React Native 0.76 on).
+# Set REACTNATIVE_MERGED_SO for libraries/apps to selectively link
```
Comment (cipolleschi) [inline packages/react-native/ReactAndroid/cmake-utils/ReactNative-application.cmake:117]: Keep the old comment please

**Review by cipolleschi — CHANGES_REQUESTED** [review]
Hi @FouadMagdy01, thank you so much to working on this problem and for trying to fix it.

The code looks good, but as a general rule, let's try to minimize the number of changes we are introducing.
In this PR there are some formatting and some comment changes that are not really needed, so it's better to revert those small changes and only keep the changes that are actually needed.

I also left a couple of questions because there are some changes that are not really clear to me.

Thank you again.

**Comment by FouadMagdy01** [comment]
Hello @cipolleschi  thank you for your time and replies
I will look into it today and commit the new code
I will also answer all of your questions 
Thank you!

File: packages/react-native/ReactAndroid/cmake-utils/ReactNative-application.cmake [hunk 10]
```diff
@@ -44,6 +45,11 @@ file(GLOB input_SRC CONFIGURE_DEPENDS
         ${REACT_ANDROID_DIR}/cmake-utils/default-app-setup/*.cpp
         ${BUILD_DIR}/generated/autolinking/src/main/jni/*.cpp)
 
+# Ensure that `input_SRC` paths use forward slashes
+foreach(path IN LISTS input_SRC)
+    string(REPLACE "\\" "/" path "${path}")
+endforeach()
+
```
Comment (FouadMagdy01) [inline packages/react-native/ReactAndroid/cmake-utils/ReactNative-application.cmake:52]: - Problem: The variables REACT_ANDROID_DIR and BUILD_DIR might also contain backslashes if they are set on a Windows system. These backslashes need to be replaced to avoid similar issues.

- Solution: Check if the script is running on Windows using if(CMAKE_HOST_WIN32). If it is, replace backslashes with forward slashes for these variables as well:

**Comment by FouadMagdy01** [comment]
@cipolleschi  Can you look into these changes now?

and i am sorry that the comments  was removed unintentionally in my first commit

this PR is tested on Android running on Windows 11
